### PR TITLE
Sharing Data between Docker Containers

### LINKED ISSUE #2151: How To Share Data between Docker Containers
Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Proposed title of article
How To Share Data between Docker Containers

### Introduction paragraph (2-3 paragraphs):

Docker is a popular containerization tool used to provide software applications with a filesystem that contains everything they need to run. Using Docker containers ensures that the software will behave the same way regardless of where it is deployed because its run-time environment is consistent.

Docker containers are transient, running just as long as it takes for the command issued in the container to complete. Sometimes, however, applications need to share access to data or persist data after a container is deleted. Persistent access to data is provided with Docker Volumes.

In this article, we will look at three different ways to share data between docker containers.

### Key takeaways: 

1. What docker container is
2. Create an independent volume
3. Create a volume from an existing directory with data
4. Share data between multiple docker containers

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by OkothPius**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
 This fixes #2151      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

File: content/articles/sharing-data-between-docker-containers/index.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
+### Sharing Data between Docker Containers
```
Comment (WanjaMIKE) [inline content/articles/sharing-data-between-docker-containers/index.md:1]: No need to add hero on this file.

File: content/articles/sharing-data-between-docker-containers/index.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+### Prerequisites
+To get started with Docker, we have to install it in our system.
+Check out this amazing [article](https://www.section.io/engineering-education/getting-started-with-docker/) on installing and getting up running with Docker. If you're not using Ubuntu be sure to check the official Docker [documentation](https://docs.docker.com/engine/install/) on how to do the same on your Operating System environment.
+We will be using Ubuntu 20.04 for this tutorial but any environment is okay.
+
+### Creating an Independent Volume
+Before we get into the creation of volumes let's get to know what they are. Docker volumes are system files that are attached to Docker containers to keep data created by running the container. Docker volumes come into play mainly when one wants to share volume among different containers, separate volumes from storage just to mention a few.
```
Comment (WanjaMIKE) [inline content/articles/sharing-data-between-docker-containers/index.md:14]: add a link to docker volumes to help the reader learn more

File: content/articles/sharing-data-between-docker-containers/index.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    {
+        "CreatedAt": "2021-05-12T20:20:24+03:00",
+        "Driver": "local",
+        "Labels": {},
+        "Mountpoint": "/var/lib/docker/volumes/NewVol1/_data",
+        "Name": "NewVol1",
+        "Options": {},
+        "Scope": "local"
+    }
+]
```
Comment (WanjaMIKE) [inline content/articles/sharing-data-between-docker-containers/index.md:69]: Please explain the above output to the reader

File: content/articles/sharing-data-between-docker-containers/index.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+```bash
+docker rm Container3 Container4 Container5
+docker volume rm NewVol3
+```
+
+So far, we've seen how data sharing happens between containers and how data volumes can be mounted as read-only.
+
+### Conclusion
+In this chapter, you went about how to create an independent docker volume that permits data to recur when a container is deleted, shared data across containers which had a caveat of the application being designed in a way to handle file locking to curb corruption of data. Ultimately, we mounted a shared volume as read-only.
```
Comment (WanjaMIKE) [inline content/articles/sharing-data-between-docker-containers/index.md:201]: ```In this article, you have learned how ...```

**Review by WanjaMIKE — CHANGES_REQUESTED** [review]
Please check out the above comments. Thanks

**Comment by OkothPius** [comment]
@WanjaMIKE I've made the changes. Thanks

**Comment by OkothPius** [comment]
@WanjaMIKE hello, any further action? Thanks

**Comment by OkothPius** [comment]
@hectorkambow hello, my reviewer seems to have been off for a while. Is there a way to know the progress on the PR? Thanks.

**Comment by WanjaMIKE** [comment]
@OkothPius - Check the message below. If you have already done so, you can ignore. Otherwise try to make a final `push` after updating your folder structure

```As you know, we were working on a new structure for our EngEd Repo, which has led to a slight slow down of article publishing. Everything is almost ready now.
In order to publish your article as soon as possible, please do the following:
- Download the most recent files off this PR (Make sure to pick the latest one with my revisions and edits).
- Delete your article directory from the articles folders.
- Do a git pull (take a look at the new structure of the repo after a git pull).```

**Comment by OkothPius** [comment]
@WanjaMIKE already did. Thanks

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by hectorkambow** [comment]
Plagiarism check done 💯

**Comment by WanjaMIKE** [comment]
Kindly, add a hero page to this artice

**Comment by OkothPius** [comment]
@WanjaMIKE done!

File: content/articles/sharing-data-between-docker-containers/index.md (hunk 1 above)
Comment (OkothPius) [reply content/articles/sharing-data-between-docker-containers/index.md:1]: You earlier said I remove it though @WanjaMIKE

**Comment by hectorkambow** [comment]
@OkothPius great work gents 

cc @WanjaMIKE 🚀 (very minor formatting edits - please go through my final edits). 👍

**Pull request merged**